crates/memory/src/op_store/time_slice_hash_store.rs

```diff
@@ -0,0 +1,532 @@
+pub enum SliceHash {
+    Single { id: u64, hash: bytes::Bytes },
+    Block { id: u64, count: u64 },
+}
+
+#[cfg(test)]
+impl SliceHash {
+    pub(crate) fn id(&self) -> u64 {
+        match self {
+            SliceHash::Single { id, .. } => *id,
+            SliceHash::Block { id, .. } => *id,
+        }
+    }
+
+    pub(crate) fn end(&self) -> u64 {
+        self.id()
+            + match self {
+                SliceHash::Single { .. } => 0,
+                SliceHash::Block { count, .. } => *count,
+            }
+    }
+}
+
+impl TimeSliceHashStore {
+    pub(super) fn new() -> Self {
+        Self {
+            highest_stored_id: None,
+            inner: vec![SliceHash::Block {
+                id: 0,
+                count: u64::MAX,
+            }],
+        }
+    }
+
+    /// Insert a hash at the given slice id.
+    pub(super) fn insert(
+        &mut self,
+        slice_id: u64,
+        hash: bytes::Bytes,
+    ) -> K2Result<()> {
+        if !hash.is_empty() {
+            match &mut self.highest_stored_id {
+                Some(id) if slice_id > *id => {
+                    *id = slice_id;
+                }
+                None => {
+                    self.highest_stored_id = Some(slice_id);
+                }
+                _ => {}
+            }
+        }
+
+        let mut handled = false;
+        for (i, slice_hash) in self.inner.iter().cloned().enumerate() {
+            match slice_hash {
+                SliceHash::Single { id, .. } => {
+                    if id == slice_id {
+                        if hash.is_empty() {
+                            // This doesn't need to be supported. If we receive an empty hash
+                            // for a slice id after we've already stored a non-empty hash for
+                            // that slice id, then the caller has done something wrong.
+
+                            return Err(K2Error::other("Cannot overwrite non-empty hash with empty hash"));
+                        } else {
+                            // We're overwriting any existing value with a new non-empty value.
+                            if let SliceHash::Single {
+                                hash: existing_hash,
+                                ..
+                            } = &mut self.inner[i]
+                            {
+                                *existing_hash = hash;
+                            }
+                        }
+
+                        handled = true;
+                        break;
+                    }
+                }
+                SliceHash::Block { id, count, .. } => {
+                    if id <= slice_id && slice_id <= id + count {
+                        // Completely contained within a compressed block then either
+                        //   - It is empty, so we can just ignore it.
+                        //   - It is not empty, so we need to split the block into two.
+
+                        if !hash.is_empty() {
```

Comment:
should this check be moved to the top of the function and always error since there is no case where an empty hash would be used?

Reply:
Hmm, that's a valid operation. If you have received no data in a time window, then when you go to compute the combined hash for it you will get no ops from the host and compute an empty hash. Given that we know it's a no-op though, it's wasteful to call the host and ask them to filter it out.

Cool, I'll make that an error

Reply:
Done: [86b826a](https://github.com/holochain/kitsune2/pull/37/commits/86b826a77b012baffaea258b7fdaa2100aef1eda)

Reply:
Never mind, due to the comment above about sparse storage, this is now gone